### PR TITLE
Check config for issues

### LINKED ISSUE #99: [BUG] GPU ids are not checked when using cfg.yaml
### 🐛 Bug

Starting a new experiment from `cfg.yaml` causes an error if number of gpu specified in `cfg.yaml` exceeds the number of gpus on the target machine.
Within the UI, this issue is not noticeable, as one can only select/deselect available GPUs.

### To Reproduce

- Use e.g. [h2ogpt-gm-oasst1-en-2048-open-llama-7b-preview-300bt](https://huggingface.co/h2oai/h2ogpt-gm-oasst1-en-2048-open-llama-7b-preview-300bt/tree/main) config that specifies 4 GPUs.
- Start a new experiment on a machine with < 4 GPUs.

```
/home/maxjeblick/.local/share/virtualenvs/h2o-llmstudio-07dpqO7E/lib/python3.10/site-packages/bitsandbytes/cuda_setup/main.py:145: UserWarning: WARNING: No GPU detected! Check your CUDA paths. Proceeding to load CPU-only library...
  warn(msg)
CUDA SETUP: Loading binary /home/maxjeblick/.local/share/virtualenvs/h2o-llmstudio-07dpqO7E/lib/python3.10/site-packages/bitsandbytes/libbitsandbytes_cpu.so...
2023-05-08 07:26:57,438 - INFO: Added key: store_based_barrier_key:1 to store for rank: 3
2023-05-08 07:26:57,472 - INFO: Added key: store_based_barrier_key:1 to store for rank: 1
2023-05-08 07:26:57,477 - INFO: Added key: store_based_barrier_key:1 to store for rank: 2
2023-05-08 07:26:57,478 - INFO: Added key: store_based_barrier_key:1 to store for rank: 0
2023-05-08 07:26:57,478 - INFO: Rank 0: Completed store-based barrier for key:store_based_barrier_key:1 with 4 nodes.
2023-05-08 07:26:57,479 - INFO: Rank 3: Completed store-based barrier for key:store_based_barrier_key:1 with 4 nodes.
2023-05-08 07:26:57,482 - INFO: Rank 1: Completed store-based barrier for key:store_based_barrier_key:1 with 4 nodes.
2023-05-08 07:26:57,487 - INFO: Rank 2: Completed store-based barrier for key:store_based_barrier_key:1 with 4 nodes.
2023-05-08 07:26:57,502 - INFO: Added key: store_based_barrier_key:2 to store for rank: 0
2023-05-08 07:26:57,511 - INFO: Added key: store_based_barrier_key:2 to store for rank: 1
2023-05-08 07:26:57,522 - INFO: Added key: store_based_barrier_key:2 to store for rank: 3
2023-05-08 07:26:57,522 - INFO: Added key: store_based_barrier_key:2 to store for rank: 2
2023-05-08 07:26:57,522 - INFO: Rank 3: Completed store-based barrier for key:store_based_barrier_key:2 with 4 nodes.
2023-05-08 07:26:57,522 - INFO: Rank 2: Completed store-based barrier for key:store_based_barrier_key:2 with 4 nodes.
2023-05-08 07:26:57,522 - INFO: Training in distributed mode with multiple processes, 1 GPU per process. Process 2, total: 4 local rank: 2.
2023-05-08 07:26:57,522 - INFO: Rank 0: Completed store-based barrier for key:store_based_barrier_key:2 with 4 nodes.
2023-05-08 07:26:57,522 - INFO: Training in distributed mode with multiple processes, 1 GPU per process. Process 0, total: 4 local rank: 0.
2023-05-08 07:26:57,523 - ERROR: Exception occurred during H2O LLM Studio run:
Traceback (most recent call last):
  File "/data/maxjeblick/PyCharmProjects/h2o-llmstudio/train_wave.py", line 106, in <module>
    run(cfg=cfg)
  File "/data/maxjeblick/PyCharmProjects/h2o-llmstudio/train.py", line 455, in run
    torch.cuda.set_device(cfg.environment._rank)
  File "/home/maxjeblick/.local/share/virtualenvs/h2o-llmstudio-07dpqO7E/lib/python3.10/site-packages/torch/cuda/__init__.py", line 350, in set_device
    torch._C._cuda_setDevice(device)
RuntimeError: CUDA error: invalid device ordinal
```

### LINKED ISSUE #190: [CODE IMPROVEMENT] add update method to ConfigProblemBase
### 🔧 Proposed code refactoring

Add a function/method that is able to automatically update `ConfigProblemBase` and its corresponding sub config classes.
It should also be able to validate if a configuration has conflicting settings.

### Motivation

1. Ability to automatically change config items (visibility/values) based on current setting in the UI. As an example, when switching backbone, automatically populate lora target modules. Also, for RLHF some parameter should be hidden, such as _disabling lora_ or some generation settings.
2. Move config changes to a single class that can be called at beginning of training. Gives a unified view on how config is changed during training. One could also enforce to only use "correct" configurations when using command line.

### LINKED ISSUE #432: [BUG] Intermediate dropout and weight decay settings seem to have no effect
### 🐛 Bug

Intermediate dropout and weight decay settings seem to have no impact on models (at least for CodeLlama models)

### To Reproduce

I've tested almost all available hyperparameters for non-rlhf training, and all have observable impacts on the model (as evidenced by perplexity and other measures of downstream performance), except for intermediate dropout and weight decay. In the debugger, I can see intermediate dropout setting `hidden_dropout_prob` and `attention_probs_dropout_prop` in `update_backbone_config`, and I can see `weight_decay` being set in `get_optimizer`, however regardless of whether I set intermediate dropout to 0 or 0.5, and weight decay at 0 or 0.5, there is no change in the perplexity after model training.

### LLM Studio version
Master latest commit

### PR DESCRIPTION
**Pull request by maxjeblick**
This PR adds checks for wrong (e.g. no gpu selected) or mutually incompatible settings.
If the configuration check fails, the experiment will not be started but a pop-up window will inform the user about the misconfiguration and how to fix it.

Also checks if the architecture has intermediate_dropout, see #432. This check is not done before the experiment has started to avoid any potential errors when calling `AutoConfig.from_pretrained` within the wave thread.

Fixes #99
(Partly) fixes #190
Fixes #432

### CODE REVIEW

File: llm_studio/python_configs/cfg_checks.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    ):
+        errors["title"] += ["Quantization without pretrained weights."]
+        errors["message"] += [
+            "Quantization is only supported for pretrained models. "
+            "Please enable pretrained model or disable quantization."
+        ]
+
+    if not cfg.training.lora and cfg.architecture.backbone_dtype != "float32":
+        if cfg.environment.mixed_precision:
+            errors["title"] += ["Mixed precision disabled."]
```
Comment (leecming82) [inline llm_studio/python_configs/cfg_checks.py:88]: Wording issue but consider changing the title for this as it gives the impression that mixed precision being disabled is the problem.

File: llm_studio/python_configs/cfg_checks.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            errors["message"] += [
+                "When not using LORA, please ensure that mixed precision is disabled "
+                "or set dtype to float32."
+            ]
+        if cfg.architecture.backbone_dtype != "bfloat16":
+            errors["title"] += ["Pure float16 or int8 training."]
+            errors["message"] += [
+                "When not using LORA, pure float16 or int8 training will "
```
Comment (leecming82) [inline llm_studio/python_configs/cfg_checks.py:96]: Include int4 here?

File: llm_studio/src/utils/modeling_utils.py [hunk 3]
```diff
@@ -483,8 +483,19 @@ def save_predictions(cfg, val_data, val_dataloader, val_df, mode):
 
 
 def update_backbone_config(config: Any, cfg: Any):
-    config.hidden_dropout_prob = cfg.architecture.intermediate_dropout
-    config.attention_probs_dropout_prob = cfg.architecture.intermediate_dropout
+    if hasattr(config, "hidden_dropout_prob"):
+        config.hidden_dropout_prob = cfg.architecture.intermediate_dropout
```
Comment (leecming82) [inline llm_studio/src/utils/modeling_utils.py:487]: Up to you - but consider logging these settings changes (before and after?)

**Review by leecming82 — APPROVED** [review]
Minor comments on messaging but otherwise LGTM!

**Comment by maxjeblick** [comment]
Thanks for the helpful comments!
I've changed the wording and added another check (sampling needs to be enabled when T>0).

**Comment by leecming82** [comment]
Thx, only outstanding nitpick is that do_sample should be turned on by default since T defaults to 0.3

**Pull request merged**